### PR TITLE
Emit warnings for certain run events in cloud backend

### PR DESCRIPTION
**Pull request by glennsarti**
The cloud backend, which communicates with TFC like APIs, can create
runs which may have one or more configuration parameters altered. These
alterations are emitted as run-events on the run so that API clients
can consume and display them to users. This commit adds a step in
plan operation to query the run-events once a run is created and then
emit specific run-event descriptions to the console as warnings for
the user.

[Jira](https://hashicorp.atlassian.net/browse/TF-5641)
[RFC](https://hashidocs.hashicorp.services/document/18gCqJdevTjMfbOfqosO8hTI90_1_WFbWIlcu32GWfVw)

## Target Release

1.4.x

### ENHANCEMENTS

- Shows warnings generated when creating a run in the cloud backend

##  Example Output

image

### CODE REVIEW

File: internal/cloud/backend_run_warning.go
Comment (uturunku1) [inline internal/cloud/backend_run_warning.go:32]: In which cases one of those actions could be set but its description would be empty?

**Review by uturunku1 — APPROVED** [review]
This looks great, Glenn. This change is going be super handy for the future.  
The output looks very clean. Though I am wondering if there could a time when the amount of warning lines we output could get a little loud for the user. Unlike the Diag warnings coming a config, I don't think the user would have an easy way to silence these warnings, unless they upgrade, no?

File: internal/cloud/backend_run_warning.go
Comment (glennsarti) [reply internal/cloud/backend_run_warning.go:11]: Yeah they can.

File: internal/cloud/backend_run_warning.go
Comment (glennsarti) [reply internal/cloud/backend_run_warning.go:32]: The backend API _shouldn't_  emit these events with an empty description but I figured there's no harm in putting in the check and the UX of `Warning: /n` is pretty bad.

**Comment by glennsarti** [comment]
> I don't think the user would have an easy way to silence these warnings, unless they upgrade, no?

Correct, but this also matches the UX on the TFC web side as well. The flip side is, if we did silence the warning then users could be confused why their policies aren't being checked or runs aren't being blocked with failed run tasks.

So I guess the question is, should these be diagnostic warnings? It's hard to say, they're not really diagnostic information, they're platform messages. But I'll defer to the maintainers on this.

File: internal/cloud/backend_run_warning.go
Comment (glennsarti) [reply internal/cloud/backend_run_warning.go:11]: Done.

**Comment by glennsarti** [comment]
> Though I am wondering if there could a time when the amount of warning lines we output could get a little loud for the user. 

Maybe? Although these particular warning only appear when you hit an **exceptional** scenario which will be rare. It's rare that you'll hit them.

**Comment by uturunku1** [comment]
> > I don't think the user would have an easy way to silence these warnings, unless they upgrade, no?
> 
> Correct, but this also matches the UX on the TFC web side as well. The flip side is, if we did silence the warning then users could be confused why their policies aren't being checked or runs aren't being blocked with failed run tasks.
> 
> So I guess the question is, should these be diagnostic warnings? It's hard to say, they're not really diagnostic information, they're platform messages. But I'll defer to the maintainers on this.
> 
> (EDIT)
> 
> So I tried some things and I don't have access to any tfdiags at the point in the code. Although I noticed that I should be using `b.CLI.Warn(....` instead of `b.CLI.Output(....`

I think you may be able to make diags available within backend_plan.go if you try something like this:
- Init diags with: `var diags tfdiags.Diagnostics`
- The warning entries returned by renderRunWarnings can get appended to diags with `tfdiags.Sourceless()`. You'd want to use tfdiags.Sourceless because your warnings do not come from a config file and so there is no code line to reference to.
- Use the available *backend.Operation (the op argument) to initialize a *backend.RunningOperation, `runningOperation, err := b.Operation(context.Background(), op)`
- Then before you return, you can report diag results with `op.ReportResult(runningOperation, diags)`

I have not tested this, so you'd have to try this out to see if actually works 😊
Also, Just FYI, that I am not suggesting using tfdiags over your original idea of just printing run event warnings with  `b.CLI.Warn()`.
This was just a recommendation in case you wanted to try it out and see how it looks.
I actually think using b.CLI.Warn is a better solution for now. Is less complicated.

**Comment by glennsarti** [comment]
FYI : @uturunku1 I don't have merge rights so I can't do much once its approved :-)

**Comment by uturunku1** [comment]
> FYI : @uturunku1 I don't have merge rights so I can't do much once its approved :-)

Gotcha! Yes, I am happy to merge this on your behalf. So you are done making changes, right? @glennsarti

**Comment by glennsarti** [comment]
> Yes, I am happy to merge this on your behalf.

Yes I am all done.  Merge away!

**Pull request merged**
